extension/flat_tensor/serialize/serialize.cpp

```diff
@@ -0,0 +1,138 @@
+    return runtime::Error::NotSupported;
+  }
+  // Create File
+  std::ofstream file;
+  file.open(path);
+  runtime::Error e = save_ptd(file, tensor_map, tensor_alignment);
+  file.close();
+  return e;
+}
+
+ET_EXPERIMENTAL runtime::Error save_ptd(
+  std::ostream& out,
+  const std::map<std::string, exec_aten::Tensor>& tensor_map,
+  const size_t tensor_alignment
+) {
+  // Assert the system is little endian
+  int n = 1;
+  if(*(char *)&n != 1) {
+    return runtime::Error::NotSupported;
+  }
+  // Create flatbuffer
+  flatbuffers::FlatBufferBuilder builder;
+
+  std::vector<flatbuffers::Offset<::flat_tensor::TensorMetadata>> tensors;
+  std::vector<flatbuffers::Offset<::flat_tensor::DataSegment>> buffers;
+
+  // Serialize the tensors.
+  size_t total_segment_size = 0;
+  for (const auto& [name, tensor] : tensor_map) {
+    auto name_offset = builder.CreateString(name);
+    // Serialize the tensor metadata.
+    auto tensor_metadata = ::flat_tensor::CreateTensorMetadata(
+      builder,
+      name_offset,
+      static_cast<executorch_flatbuffer::ScalarType>(tensor.scalar_type()),
+      builder.CreateVector(tensor.sizes().data(), tensor.sizes().size()),
+      builder.CreateVector(tensor.dim_order().data(), tensor.dim_order().size()),
+      0, // segment index
+      total_segment_size);
+    tensors.push_back(tensor_metadata);      // Precalculate the size of the data blob
+    total_segment_size += aligned_size(tensor.nbytes(), tensor_alignment);
+  }
+  // Only have one segment
+  buffers.push_back(::flat_tensor::CreateDataSegment(builder, 0, total_segment_size));
+  
+  auto flat_tensor = CreateFlatTensor(builder, internal::VERSION, tensor_alignment, builder.CreateVector(tensors), builder.CreateVector(buffers));
+  builder.Finish(flat_tensor);  // Our flatbuffer is created now.
+
+
+  // Calculate flatbuffer padding.
+  auto padded_flatbufer_size = aligned_size(builder.GetSize(), tensor_alignment);
+  auto padded_header_size = aligned_size(internal::HEADER_EXPECTED_LENGTH, tensor_alignment);
+
+  // Serialize header
+  out.write(internal::HEADER_MAGIC, sizeof(internal::HEADER_MAGIC));
+  out.write(reinterpret_cast<const char *>(&internal::HEADER_EXPECTED_LENGTH), sizeof(internal::HEADER_EXPECTED_LENGTH));
+
+  internal::FlatTensorHeader header = {
+    padded_header_size, // Offset to flatbuffer
+    builder.GetSize(), // flatbuffer size
+    padded_header_size + padded_flatbufer_size, // offset to segments
+    total_segment_size // segment data size
+  };
+
+  out.write(reinterpret_cast<const char *>(&header.flatbuffer_offset), sizeof(header.flatbuffer_offset));
+  out.write(reinterpret_cast<const char *>(&header.flatbuffer_size), sizeof(header.flatbuffer_size));
+  out.write(reinterpret_cast<const char *>(&header.segment_base_offset), sizeof(header.segment_base_offset));
+  out.write(reinterpret_cast<const char *>(&header.segment_data_size), sizeof(header.segment_data_size));
+
+  // serialize header padding
```

Comment:
Super nit: these steps aren't serializing, they're writing. Serializing is the act of arranging the data in a particular format, regardless of where that data lives. These steps are writing out already-serialized data.